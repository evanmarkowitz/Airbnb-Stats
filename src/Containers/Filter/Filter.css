.filter-section {
  display: flex;
  flex-direction: column;
  margin: auto;
  width: 50%;
  background: white;
  opacity: .98;
  text-align: center;
}

.buttons {
  display: flex;
  justify-content: space-around;
}

.filter-button {
  width: 120px;
  height: 50px;
  background-color: #00d1fb;
  margin: 10px;
  font-size: 16px;
  box-sizing: border-box;
  padding: 4px;
  font-weight: bold;
}