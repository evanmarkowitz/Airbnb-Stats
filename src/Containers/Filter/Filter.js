import React, { Component } from 'react'
import './Filter.css'
import { fetchApartments, apartmentCleaner } from '../../ApiCalls/apiCall'
import {getApts, getHood, getAptType} from '../../actions/index.js'
import { connect } from 'react-redux';

export class Filter extends Component {
  constructor() {
    super()
    this.state = {
      city: "New-york-city",
      roomType: '',
      borough: '',
      Brooklyn: ['Williamsburg', 'Bedford-Stuyvesant', 'Bushwick', 'Crown Heights', 'Greenpoint', 'Financial District', 'Flatbush',
    'Clinton Hill', 'Park Slope', 'Fort Greene', 'Sunset Park'],
      Manhattan: ['Harlem', 'Upper West Side', 'Hell/s Kitchen', 'East Village', 'Upper East Side', 
      'Midtown', 'Chelsea',  'Lower East Side', 'West Village', 'Murray Hill', 'Greenwich Village', 'Soho', 'Chinatown', 'Gramercy'], 
      Queens: ['Astoria', 'Long Island City', 'Flushing', 'Ridgewood', 'Sunnyside', 'Ditmars Steinway' ],
      chosenHood: '',
      typeOfAprtment: false,
      boroughShow: false,
      hoodsShow: false,
    }
  }

  toggleType = () => {
    this.setState({typeOfAprtment: !this.state.typeOfAprtment})
  }
  toggleBorough= () => {
    this.setState({boroughShow: !this.state.boroughShow})
  }
  toggleHood = () => {
    this.setState({hoodsShow: !this.state.hoodsShow})
  }

  handleChange = async (event) => {
    await this.setState({[event.target.name]: event.target.value})
  }

  buildNeighborhood = () => {
    let hoods =  this.state[this.state.borough] || []
    return hoods.map(hood => {
      return <button
        name='hood'
        value={hood}
        onClick={event => this.chooseHood(event)}
        className='filter-button'>{hood}
      </button>
    })
  }

  chooseHood = async (event) => {
    let chosenHood =  [event.target.value][0]
    await this.props.getHood(chosenHood)
  }

  chooseAptType = (event) => {
    console.log([event.target.value][0])
    let chosenAptType =  [event.target.value][0]
    this.props.getAptType(chosenAptType)
  }
  
  render() {
  
    return (
      <section className = 'filter-section'>
        <section className ='type-of-apartment'>
          <h3 onClick={this.toggleType}>Type of apartment</h3>

         {this.state.typeOfAprtment && 
         <div className='buttons'>
            <button onClick={(event) => this.chooseAptType(event)}
            name='roomType'
            value='Entire home/apt'
            className='filter-button'>Entire Apartment</button>

            <button onClick={(event) => this.chooseAptType(event)}
            name='roomType'
            value='Private room'
            className='filter-button'>Room</button>
          </div>}
        </section>
        <section className='borough'>
          <h3 onClick={this.toggleBorough}>Borough</h3>
          {this.state.boroughShow && 
            <div className='buttons'>
              <button className='filter-button'
              onClick={(event) => this.handleChange(event)}
              name='borough'
              value='Manhattan'
              >Manhattan</button>

              <button className='filter-button'
              onClick={(event) => this.handleChange(event)}
              name='borough'
              value='Brooklyn'>
              Brooklyn</button>

              <button className='filter-button'
              onClick={(event) => this.handleChange(event)}
              name='borough'
              value='Queens'>
              Queens</button>
            </div>
          }
        </section>
        
        <section className='neighborhoods'>
          {this.state.borough &&
          <h3 onClick={this.toggleHood}>Neighborhoods</h3>}
          
          <div className='buttons'>
            {this.buildNeighborhood()}
          </div> 
        </section> 
      </section>
    )
  }
}

export const mapStateToProps = (state) => ({
  apts: state.apts
})

export const mapDispatchToProps = (dispatch) => ({
  getApts: (apts) => dispatch(getApts(apts)),
  getHood: (hood) => dispatch(getHood(hood)),
  getAptType: room => dispatch(getAptType(room))
})

export default connect(mapStateToProps, mapDispatchToProps)(Filter)