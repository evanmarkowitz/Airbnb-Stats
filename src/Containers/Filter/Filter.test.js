import {mapStateToProps, mapDispatchToProps, Filter} from './Filter'
import { getApts, getHood, getAptType } from '../../actions/index.js'
import {shallow} from 'enzyme'
import React from 'react'


describe('Filter', () => {
  let wrapper;
  let apts;
  let getApts; 
  let getHood;
  let getAptType;
  beforeEach(() => {
    wrapper = shallow(<Filter apts ={apts} getApts={getApts} getHood={getHood} getAptType={getAptType}/>)
    apts = []
    getApts = jest.fn()
    getHood = jest.fn()
    getAptType = jest.fn()
  })

  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should toggle state when toggleHood is called', () => {
    expect(wrapper.state('hoodsShow')).toEqual(false)
    wrapper.instance().toggleHood()
    expect(wrapper.state('hoodsShow')).toEqual(true)
  })

  it('should toggle state when toggleBorough is called', () => {
    expect(wrapper.state('boroughShow')).toEqual(false)
    wrapper.instance().toggleBorough()
    expect(wrapper.state('boroughShow')).toEqual(true)
  })

  it('should toggle state when toggleType is called', () => {
    expect(wrapper.state('typeOfAprtment')).toEqual(false)
    wrapper.instance().toggleType()
    expect(wrapper.state('typeOfAprtment')).toEqual(true)
  })

  it('should update state when a neighborhood is chosen', () => {
    const mockEvent = { target: { name: 'hood', value: 'Williamsburg'}}
    const expected = 'Williamsburg'
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('hood')).toEqual(expected);
  });

  it('should update state when a borough is choose is chosen', () => {
    const mockEvent = { target: { name: 'borough', value: 'Brooklyn'}}
    const expected = 'Brooklyn'
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('borough')).toEqual(expected);
  });

  it.skip('should update state when a hood is chosen', async () => {
    const mockEvent = { target: { name: 'hood', value: 'Williamsburg'}}
    const mockEventHood = 'Williamsburg'
    wrapper.instance().getHood = jest.fn()
    await wrapper.instance().chooseHood(mockEvent)
    await expect(wrapper.instance().getHood).toHaveBeenCalled()
  });

  it('should map state to props', () => {
    let initialState = {apts: [{name:'hi'}]}
    wrapper = shallow(<Filter initialState = {initialState} getApts={getApts} getHood={getHood} getAptType={getAptType}/>)
    const mappedProps = mapStateToProps(initialState);
    expect(mappedProps).toEqual(initialState);
  })
  it('filters when the filter button is clicked', () => {
    wrapper.instance().handleChange = jest.fn()
    wrapper.setState({boroughShow: true})
    wrapper.find('.filter-man').simulate('click')
    expect(wrapper.instance().handleChange).toHaveBeenCalled()
  })
  it('filters when the filter button is clicked', () => {
    wrapper.instance().handleChange = jest.fn()
    wrapper.setState({boroughShow: true})
    wrapper.find('.filter-bk').simulate('click')
    expect(wrapper.instance().handleChange).toHaveBeenCalled()
  })
  it('filters when the filter button is clicked', () => {
    wrapper.instance().handleChange = jest.fn()
    wrapper.setState({boroughShow: true})
    wrapper.find('.filter-queens').simulate('click')
    expect(wrapper.instance().handleChange).toHaveBeenCalled()
  })
  it('filters when the filter button is clicked', () => {
    wrapper.instance().chooseAptType = jest.fn()
    wrapper.setState({typeOfAprtment: true})
    wrapper.find('.entire-home').simulate('click')
    expect(wrapper.instance().chooseAptType).toHaveBeenCalled()
  })
  it('filters when the filter button is clicked', () => {
    wrapper.instance().chooseAptType = jest.fn()
    wrapper.setState({typeOfAprtment: true})
    wrapper.find('.private-room').simulate('click')
    expect(wrapper.instance().chooseAptType).toHaveBeenCalled()
  })
  it.skip('should call getHood when chooseHood is called', () => {
    let mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().props.getHood = jest.fn()
    wrapper.instance().chooseHood(mockEvent)
    expect(wrapper.instance().props.getHood).toHaveBeenCalled()
  })
})
  describe('Filter: Map Dispatch To Props', () => {
    let initialState;
    let mockDispatch
    beforeEach(() => {
      initialState = {apts: [{name:'hi'}], hood: 'Williamsburg', aptType: 'Private Room' }
      mockDispatch = jest.fn();
    })

  it('should map getApt to props', () => {
    const mockAction = getApts(initialState.apts)
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getApts(initialState.apts);
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })

  it('should map getHood to props', () => {
    const mockAction = getHood(initialState.hood)
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getHood(initialState.hood);
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })

  it('should map getHood to props', () => {
    const mockAction = getAptType(initialState.aptType)
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.getAptType(initialState.aptType);
    expect(mockDispatch).toHaveBeenCalledWith(mockAction);
  })

})

