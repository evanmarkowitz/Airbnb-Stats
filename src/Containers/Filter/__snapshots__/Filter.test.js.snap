// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should match a snapshot 1`] = `ShallowWrapper {}`;
